### ISSUE TITLE
asyncio.run make asyncssh Connection lost strangely. bug?

### ISSUE THREAD

**Issue by NewUserHa** [ronf/asyncssh#220]
```
import asyncio
import asyncssh
import logging
class machine:
    pass
a = machine()
a.host = '192.168.0.1'
a.password = '1'
b = machine()
b.host = '192.168.0.2'
b.password = '2'
machines = [a, b]

async def connect_with_retry(machine):
    machine.conn = await asyncssh.connect(machine.host, username='root', password=machine.password, known_hosts=None, keepalive_interval=1)

async def run_command(machine, command='ll', timeout=2):
    return 1

async def main():
    for _ in machines:
        asyncio.create_task(connect_with_retry(_))
    # await asyncio.sleep(5)
    print(await asyncio.gather(*[run_command(_, 'll')for _ in machines]))

loop = asyncio.get_event_loop()
asyncio.get_event_loop().set_debug(True)
logging.basicConfig(level=logging.DEBUG)
asyncio.run(main(), debug=True)
```
will always throw `asyncssh.misc.ConnectionLost: Connection lost`,
  except using `run_until_completed` or `run_forever` or `uncomment asyncio.sleep()`

In ./Lib/asyncio/runners.py, the `run` definetion is 
```
def run(main, *, debug=False):
...
    try:
        events.set_event_loop(loop)
        loop.set_debug(debug)
        return loop.run_until_complete(main)
    finally:
        try:
            _cancel_all_tasks(loop)
            loop.run_until_complete(loop.shutdown_asyncgens())
...

def _cancel_all_tasks(loop):
    to_cancel = tasks.all_tasks(loop)
    if not to_cancel:
        return

    for task in to_cancel:
        task.cancel()

    loop.run_until_complete(
        tasks.gather(*to_cancel, loop=loop, return_exceptions=True))
```
it does cancel while asyncssh connecting. let's assume asyncio is right, then is it a bug since it's strange and taking time to find out what's going on?

**Comment by ronf** [comment]
One thing that immediately jumps out at me here is that you are not waiting for the connect_with_retry() calls to finish before you attempt to perform the run_command() actions. In the code here, run_command() just does a "return 1" at the moment, but I expect in any real version of this code, it would attempt to use machine.conn, but in most cases that probably wouldn't be set at the time run_command() tries to use it.

Rather than using create_task() here on the calls to connect_with_retry(), I think you're going to want to do a gather() on them like what you're doing with run_command(). Alternately, you could have a single function that does the connect and runs commands, and that function could be what you do the gather() on.

**Comment by NewUserHa** [comment]
`returning 1 is` demo code to make the issue as simple as possible to say 'what in `run_command` is unimportant'

the real case there is 'it will prepare all connection and send command by web page submitting.' so I cant `await connect_with_retry(..)` or `gather` and connections have to be prepared before all other ssh commands
I guess it may need asyncio.lock() to achieve this.

because weather `run command()` is calling/awaiting `machine.conn.run(..)` will cause this.
if assuming asyncio.run() is all correct, should asyncssh handle this situation?

**Comment by ronf** [comment]
It looks to me like both asyncio and asyncssh are behaving as expected here. When main() exits, any active tasks started by main() will be canceled. What behavior were you looking for?

If you need to wait until an SSH connection is opened before trying to run commands on it, I would recommend the use of an asyncio.Event object in your "machine" objects. You can create the Event() when you first populate the machine objects, and have the connect_with_retry() call set() on the Event just after filling in machine.conn. Then, in run_command(), you can do an await on wait() on the Event. This will return immediately once the event is set, but asynchronously block if the SSH handshake hasn't completed yet. By the time wait() returns, the machine.conn should be initialized and safe to use.

If you do this, you shouldn't need to gather() or otherwise wait for the connections to complete.

On another note, according to the documentation for asyncio.run(), it always creates a new event loop. So, the following lines don't really do anything:

```python
loop = asyncio.get_event_loop()
asyncio.get_event_loop().set_debug(True)
```

Passing debug=True to the run() call will achieve the same result, though, and you're already doing that.

**Comment by NewUserHa** [comment]
Thanks so much for your amazing advice! It always helps a lot. I'll try `asyncio.Event` ASAP.

it's some weird to me, a user who is not familiar with detailed implements of ssh, throw `Connection lost` before `Connection made`. I feel it could throw `Connection canceled` but should it be the job of asyncssh? maybe be of user code?

I'm now trying to do better on programming design.

**Comment by NewUserHa** [comment]
this is the `connect_with_retry` write by me, https://github.com/ronf/asyncssh/issues/212 `doing a conn.close() in the "finally"` may be wrong because conn may not exist if `await asyncssh.connect` thorw error or from wait_for.
<br /><br />
```
import asyncio
import asyncssh
import logging
import sys


class machine:
    async def connect_with_retry(self):
        times = 1
        while times <= 3:
            try:
                self.conn = await asyncio.wait_for(asyncssh.connect(self.host, username='root', password=self.password, known_hosts=None, keepalive_interval=0), 100)
                await self.conn.wait_closed()
                logging.error(f'tried {self.host} {times} times')
                await asyncio.sleep(2)
            except asyncio.CancelledError:
                logging.info(f'{self.host} CANCELED~')
                # break
            except (asyncio.TimeoutError, Exception) as e:
                self.status = str(e)
            times += 1

    async def run_command(self, command='', timeout=2):
        try:
            cp = await asyncio.wait_for(self.conn.run(command), timeout)
            result = (cp.stdout, cp.stderr)
        except asyncio.TimeoutError:
            result = ('timeout',) * 2
        except AttributeError:
            result = ('unavailable',) * 2
        return result


a = machine()
a.host = '192.168.0.234'
a.password = '123456-234'
b = machine()
b.host = '192.168.0.235'
b.password = '123456-235'
machines = [a, b]


async def main():
    for _ in machines:
        asyncio.create_task(_.connect_with_retry())
    await asyncio.sleep(3) # this line will not stop cancel but only delay it.
    # print(await asyncio.gather(*[_.run_command('ll')for _ in machines])) #without delay, this line will cause two CANCELED logs first then one or two Connection lost exception and make connected `conn=` number start from 1 or 2. with delay, no exception, two CANCELED logs, connected `conn=` start from 2.
    print(await asyncio.gather(*[run_command(_, 'll')for _ in machines])) #without delay, this line will cause two CANCELED logs first then no Connection lost, no no `run_command` exception as excepted, and connected `conn=` number start from 0. with delay, two connected first then 2 CANCELED logs, then connected `conn=` start from 2.

logging.basicConfig(level=logging.DEBUG)
asyncio.run(main(), debug=True)
```
All three line with comments's behavior are understandable except `why does connection lost diappear in 3rd` and `why conn= number is not consistent`. And shall the canceled asyncssh.connect task throw `Connection lost` after cathed `Cancelederror`?

If I have hundreds of `connect_with_retry()` then I need to create hundreds of `asyncio.Event` objects? how about using `machine` object's `conn` attribute to implement like `asyncio.Event` using `loop.create_future()` and `fut.set_result(True)`?

**Comment by ronf** [comment]
> this is the connect_with_retry write by me, #212 doing a conn.close() in the "finally" may be wrong because conn may not exist if await asyncssh.connect thorw error or from wait_for.

If you wanted to do something like a "conn.close()" in a "finally" clause, you'd want to initialize "conn" to None and later test in the "finally" if conn was set or not before calling "close".

> All three line with comments's behavior are understandable except why does connection lost diappear in 3rd and why conn= number is not consistent. And shall the canceled asyncssh.connect task throw Connection lost after cathed Cancelederror?

I'm not sure what you mean about the number being not consistent. As for the error thrown, that will depend on how far things get before you do the cancel. If you are still waiting for the TCP connection to complete at the time of the cancel, the error returned would be up to  asyncio's create_connection() to decide what to return. However, if the TCP connection was opened successfully and it was in the middle of an SSH handshake, then you'd get a ConnectionLost error. In that case, it would also have created your SSHClientConnection object and called connection_made() on it before triggering the connection_lost().

> If I have hundreds of connect_with_retry() then I need to create hundreds of asyncio.Event objects? how about using machine object's conn attribute to implement like asyncio.Event using loop.create_future() and fut.set_result(True)?

The benefit to using an Event here is that you can potentially wait on it multiple times. Consider a case where you have hundreds of requests, but some of those requests are going to the same server. You have only once machine object per server and should be able to call connect_with_retry() only once for each, but once that server connection is open you could call run_command() on it multiple times, with each command getting its own independent session on a shared connection. Each time you want to run a command, you'd wait on the Event associated with the "machine" object associated with the server you want to run the command on to wait until the connection is established, and you might have multiple of these waits happening in parallel, but they'd all unblock as soon as set() is called on the Event (when the handshake completes).

I don't think there's much difference in cost between an Event object and a Future object (certainly not compared to all the other work being done to complete an SSH handshake), so I wouldn't worry about that.

**Comment by ronf** [comment]
> it's some weird to me, a user who is not familiar with detailed implements of ssh, throw Connection lost before Connection made. I feel it could throw Connection canceled but should it be the job of asyncssh? maybe be of user code?

I did some experimentation today and wanted to check with you on something. Are you actually seeing cases where you get back a ConnectionLost exception when asyncio.run() cancels one of your tasks? I haven't been able to reproduce that here. Whether the cancel happens during the TCP connect, the SSH handshake, or after that when doing something like a read() on an open SSH channel, the exception raised is always CancelledError() in the tests I ran.  If you can reproduce a case where exiting from asyncio.run() results in some other exception, I'd love to take a closer look.

Note: I would expect you to get ConnectionLost back on something like a task doing a read() on an open SSH channel if you called close() explicitly on the connection while the read was outstanding, but you ConnectionLost shouldn't be raised if the connection was never successfully opened. Similarly, if you have your own SSHConnection subclass, you should not see the connection_lost() callback called on in cases where connection_made() was never called.

**Comment by NewUserHa** [comment]
the code I posted is the full code. using
```
async def main():
    for _ in machines:
        asyncio.create_task(_.connect_with_retry())
    print(await asyncio.gather(*[_.run_command('ll')for _ in machines]))
```
can always reproduce:(but the except can appear once or twice depend on the time depended on its progress I think.)
```
DEBUG:asyncio:Using selector: SelectSelector
INFO:asyncssh:Opening SSH connection to 192.168.0.2, port 22
DEBUG:asyncio:connect <socket.socket fd=556, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6> to ('192.168.0.2', 22)
INFO:asyncssh:Opening SSH connection to 192.168.0.3, port 22
DEBUG:asyncio:connect <socket.socket fd=768, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6> to ('192.168.0.3', 22)
[('unavailable', 'unavailable'), ('unavailable', 'unavailable')]
INFO:asyncssh:[conn=0] Connection to 192.168.0.3, port 22 succeeded
INFO:asyncssh:[conn=0]   Local address: 192.168.0.2, port 63529
INFO:asyncssh:[conn=1] Connection to 192.168.0.2, port 22 succeeded
INFO:asyncssh:[conn=1]   Local address: 192.168.0.2, port 63528
INFO:root:192.168.0.3 CANCELED~
INFO:root:192.168.0.2 CANCELED~
INFO:asyncssh:Opening SSH connection to 192.168.0.3, port 22
DEBUG:asyncio:connect <socket.socket fd=768, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6> to ('192.168.0.3', 22)
INFO:asyncssh:Opening SSH connection to 192.168.0.2, port 22
DEBUG:asyncio:connect <socket.socket fd=556, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6> to ('192.168.0.2', 22)
ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=ConnectionLost('Connection lost') created at C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py:312>
source_traceback: Object created at (most recent call last):
  File "C:\Users\kakik\Desktop\a.py", line 51, in <module>
    asyncio.run(main(), debug=True)
  File "C:\Program Files\Python37\lib\asyncio\runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 571, in run_until_complete
    self.run_forever()
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 539, in run_forever
    self._run_once()
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 1767, in _run_once
    handle._run()
  File "C:\Program Files\Python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Program Files\Python37\lib\asyncio\tasks.py", line 603, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\misc.py", line 188, in __await__
    return (yield from self._coro)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5724, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5332, in create_connection
    local_addr=local_addr)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 986, in create_connection
    ssl_handshake_timeout=ssl_handshake_timeout)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 1002, in _create_connection_transport
    protocol = protocol_factory()
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5241, in conn_factory
    agent_path)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 2310, in __init__
    keepalive_count_max, server=False)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 312, in __init__
    self._kex_waiter = asyncio.Future(loop=loop)
asyncssh.misc.ConnectionLost: Connection lost
INFO:asyncssh:[conn=1] Connection lost
ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=ConnectionLost('Connection lost') created at C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py:312>
source_traceback: Object created at (most recent call last):
  File "C:\Users\kakik\Desktop\a.py", line 51, in <module>
    asyncio.run(main(), debug=True)
  File "C:\Program Files\Python37\lib\asyncio\runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 571, in run_until_complete
    self.run_forever()
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 539, in run_forever
    self._run_once()
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 1767, in _run_once
    handle._run()
  File "C:\Program Files\Python37\lib\asyncio\events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Program Files\Python37\lib\asyncio\tasks.py", line 603, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\misc.py", line 188, in __await__
    return (yield from self._coro)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5724, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5332, in create_connection
    local_addr=local_addr)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 986, in create_connection
    ssl_handshake_timeout=ssl_handshake_timeout)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 1002, in _create_connection_transport
    protocol = protocol_factory()
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5241, in conn_factory
    agent_path)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 2310, in __init__
    keepalive_count_max, server=False)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 312, in __init__
    self._kex_waiter = asyncio.Future(loop=loop)
asyncssh.misc.ConnectionLost: Connection lost
INFO:asyncssh:[conn=0] Connection lost
INFO:asyncssh:[conn=2] Connection to 192.168.0.3, port 22 succeeded
INFO:asyncssh:[conn=2]   Local address: 192.168.0.2, port 63530
DEBUG:asyncio:<socket.socket fd=768, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.0.2', 63530), raddr=('192.168.0.3', 22)> connected to 192.168.0.3:22: (<_SelectorSocketTransport fd=768 read=polling write=<idle, bufsize=0>>, <asyncssh.connection.SSHClientConnection object at 0x000001C9D9E53B70>)
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
INFO:asyncssh:[conn=3] Connection to 192.168.0.2, port 22 succeeded
INFO:asyncssh:[conn=3]   Local address: 192.168.0.2, port 63531
DEBUG:asyncio:<socket.socket fd=556, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('192.168.0.2', 63531), raddr=('192.168.0.2', 22)> connected to 192.168.0.2:22: (<_SelectorSocketTransport fd=556 read=polling write=<idle, bufsize=0>>, <asyncssh.connection.SSHClientConnection object at 0x000001C9DA37BCC0>)
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
DEBUG:asyncssh:[conn=2] Requesting key exchange
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
DEBUG:asyncssh:[conn=2] Received key exchange request
DEBUG:asyncssh:[conn=2] Beginning key exchange
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
DEBUG:asyncssh:[conn=3] Requesting key exchange
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
DEBUG:asyncssh:[conn=3] Received key exchange request
DEBUG:asyncssh:[conn=3] Beginning key exchange
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
DEBUG:asyncssh:[conn=2] Completed key exchange
DEBUG:asyncio:poll 99985.000 ms took 0.000 ms: 1 events
INFO:asyncssh:[conn=2] Beginning auth for user root
DEBUG:asyncssh:[conn=3] Completed key exchange
DEBUG:asyncio:poll 99969.000 ms took 0.000 ms: 1 events
INFO:asyncssh:[conn=3] Beginning auth for user root
DEBUG:asyncssh:[conn=2] Trying password auth
DEBUG:asyncssh:[conn=3] Trying password auth
DEBUG:asyncio:poll 99969.000 ms took 16.000 ms: 1 events
INFO:asyncssh:[conn=2] Auth for user root succeeded
DEBUG:asyncio:poll 99953.000 ms took 0.000 ms: 1 events
INFO:asyncssh:[conn=3] Auth for user root succeeded
DEBUG:asyncio:poll took 109.000 ms: 1 events
DEBUG:asyncssh:[conn=2] Received unknown global request: hostkeys-00@openssh.com
DEBUG:asyncio:poll took 0.000 ms: 1 events
DEBUG:asyncssh:[conn=3] Received unknown global request: hostkeys-00@openssh.com
...hang here...
```

**Comment by ronf** [comment]
Ok -  I haven't figured out all of the details yet, but one issue is that you are catching the "CancelledError" in your connect_with_retry() and looping back around to make a second connection when asyncio.run() attempts to cancel all the tasks. This allows the second set of connections to succeed even though the first round was canceled. These new connections never break out of the wait_closed() call, though, as nothing is being done to close those connections once they are open, which is why the program seems to "hang". I'm not sure yet what's going on with the exception that was never retrieved, but I suspect it may have something to do with the "conn" member being overwritten and the garbage collector trying to clean up the first set of connections that were canceled before their handshake completed.

You're definitely going to want to avoid letting main() exit (and thus cancel all the active tasks) until you really want the program to exit. As it is right now, the CancelledError triggered by that is being misinterpreted as the cancel you'd expect to get from the wait_for() call, and that's part of what is leading to some of the confusion here.

**Comment by NewUserHa** [comment]
codes in my first comment are the simplest one to reproduce `Connect lost` exception.
`ERROR:asyncio:Future exception was never retrieved` I guess it's `Connect lost` exception since debug is true.
or `If aw is a coroutine it is automatically scheduled as a Task.` from https://docs.python.org/3/library/asyncio-task.html#asyncio.wait_for . maybe nothing is cathing `asyncio.create_task(_.connect_with_retry())`'s exception?

**Comment by ronf** [comment]
Yes - independent of the other issues here, it's a problem that nothing is looking at the result coming back from the task created to do the connect_with_retry().

It's interesting that the output says "Future exception was never retrieved" rather than "Task exception was never retrieved", though. That suggests that the call to asyncssh.connect() never got far enough to actually do the await on the _kex_waiter Future object. So, that object was created and the ConnectionLost exception was eventually set on it, but asyncssh.connect() didn't get to run wait_kex() to pick up that exception. I don't quite see how that would be possible, though -- the call to asyncio.create_connection() would have to succeed in opening the TCP connection for the ConnectionLost exception to eventually be set on _kex_waiter, and if that succeeds, asyncssh.create_connection()'s very next action is to call wait_kex() to retrieve that Future result. Perhaps the await on that Future is what is getting canceled here, but later the ConnectionLost is still getting triggered on the orphaned connection.

If you update your code to avoid cancelling that first set of connection attempts due to asyncio.run() exiting, I think this error will likely go away.

**Comment by NewUserHa** [comment]
wow, fantastic.

`avoid cancelling`, mean `shield`? with the first code, both wrapping `asyncssh.connect` or `connect_with_retry` don't work.
but, I found await in `main's for loop` can avoid the `Connection lost`.

**Comment by ronf** [comment]
I wasn't thinking of shield. In the real version of this code, I expect you wouldn't be exiting out of the asyncio.run(), as you'd have to keep the event loop running to receive web requests. As a result, I would not expect the original tasks you start to open the SSH connections to be cancelled the way they were in this test. That should eliminate the "connection lost" errors, and also avoid the second set of connections being opened due to the connect_with_retry() code (which explains the difference in numbering you saw).

**Comment by ronf** [comment]
Closing due to inactivity. If you have additional questions, feel free to open a new issue.
